### PR TITLE
Fix lab build prod

### PR DESCRIPTION
**Pull request by telamonian**
Quoting @tibdex from #6839:

## References

- https://github.com/jupyterlab/jupyterlab/issues/6804#issuecomment-510721047
- https://github.com/jupyterlab/jupyterlab/issues/4276

## Code changes

The changes are small an explained in comments and in the commit messages.

## User-facing changes

End-users will have a better UX since they will be running optimized code.

## Backwards-incompatible changes

There should be none as switching to prod build by default is non breaking and Webpack is an implementation detail.

### CODE REVIEW

**Comment by telamonian** [comment]
I added a couple of commits to this PR that:

- add `minimize` as an option for `lab build`.
- defaults the `lab build` to dev if there's anything in the `"linked_packages"` or `"local_extensions"` fields of `build_config.json`, and to prod otherwise.

**Comment by telamonian** [comment]
Quoting @blink1073:

After a rebase I think the logic looks good. Let’s wait until master is at 1.1dev to merge since this is a cli change

**Comment by telamonian** [comment]
Rebase is done, this is ready to go when 1.1dev is created.

File: yarn.lock [hunk 1]
```diff
@@ -5649,6 +5649,11 @@ fs.realpath@^1.0.0:
   resolved "https://registry.yarnpkg.com/fs.realpath/-/fs.realpath-1.0.0.tgz#1504ad2523158caa40db4a2787cb01411994ea4f"
   integrity sha1-FQStJSMVjKpA20onh8sBQRmU6k8=
 
+fs@~0.0.1-security:
+  version "0.0.1-security"
+  resolved "https://registry.yarnpkg.com/fs/-/fs-0.0.1-security.tgz#8a7bd37186b6dddf3813f23858b57ecaaf5e41d4"
+  integrity sha1-invTcYa23d84E/I4WLV+yq9eQdQ=
+
```
Comment (vidartf) [inline yarn.lock]: (non-blocking:) Didn't you fix this in https://github.com/jupyterlab/jupyterlab/pull/6889 already?

Comment (telamonian) [reply yarn.lock]: yep. The branch probably needs an integrity run

Comment (telamonian) [reply yarn.lock]: Done! Although there's still one little bothersome difference in `yarn.lock` for no apparent reason. Annoying

Comment (telamonian) [reply yarn.lock]: Annnd now that's taken care of too. Hooray!

**Comment by jasongrout** [comment]
@blink1073 - merge if you think this is ready to be merged

**Comment by blink1073** [comment]
In it goes!

**Comment by jasongrout** [comment]
Should we backport this to 1.0.x? In the end, did it break any apis that someone may be using developing an extension?

**Comment by blink1073** [comment]
We changed the default CLI behavior, which could break some builds.

**Comment by jasongrout** [comment]
So it's a *breaking* change, as in 2.0? How would a build break?

**Comment by blink1073** [comment]
Running out of memory because it would now use the Uglify plugin.

**Comment by telamonian** [comment]
There are definitely users who will have to adjust their build commands. For example, [this guy](https://github.com/jupyterlab/jupyterlab/issues/3561#issuecomment-521649529) is an example of user who would probably now need to do `jupyter lab build --minimize=False`.

It seems that the users who will run into trouble are those with the special circumstance of trying to build Jlab in a resource-constrained VM. Currently, they're essentially relying on an unintended behavior of `jupyter lab build` (it will always do a `dev` build, even for end users, and the `dev` build just so happens to not use the memory-intensive minifier). 

However, unlike a breaking change to an API, there will be an easy way for the affected end users to fix their own problem (ie by adding the `--minimize=False` arg). So I think the best solution would be to add an informative message when the build fails. The message would explain the `--dev-build` and `--minimize` args, and also explain how to add them to their config files in order to ensure that their builds always work correctly.

I doubt that we can reliably catch all flavors of out of memory errors, so we could maybe just have that message displayed whenever the build fails for whatever reason.

**Comment by vidartf** [comment]
> Running out of memory because it would now use the Uglify plugin.

*Terser plugin, right?

**Comment by blink1073** [comment]
👍 for a backport after adding the informative message.  

> *Terser plugin, right?

Yeah, that one :wink:

**Comment by telamonian** [comment]
I took a stab at adding the troubleshooting message in #7037. The PR probably isn't ready yet, and could definitely use some feedback.

**Pull request merged**